### ISSUE TITLE
Vintage engine reports intermediate containers as finished before executing dynamic children

### ISSUE THREAD

**Issue by marcphilipp** [junit-team/junit5#1819]
For the demo project below, Vintage reports the `AbcTest` class as finished before starting its dynamic children:

```
Started:     JUnit Vintage ([engine:junit-vintage])
Started:     AbcTestSuite ([engine:junit-vintage]/[runner:AbcTestSuite])
Started:     AbcTest ([engine:junit-vintage]/[runner:AbcTestSuite]/[test:AbcTest])
Started:     simple ([engine:junit-vintage]/[runner:AbcTestSuite]/[test:AbcTest]/[test:simple(AbcTest)])
Finished:    simple ([engine:junit-vintage]/[runner:AbcTestSuite]/[test:AbcTest]/[test:simple(AbcTest)])
Finished:    AbcTest ([engine:junit-vintage]/[runner:AbcTestSuite]/[test:AbcTest])
Test registered:   aaa 1 ([engine:junit-vintage]/[runner:AbcTestSuite]/[test:AbcTest]/[dynamic:aaa 1(AbcTest)])
Started:     aaa 1 ([engine:junit-vintage]/[runner:AbcTestSuite]/[test:AbcTest]/[dynamic:aaa 1(AbcTest)])
Finished:    aaa 1 ([engine:junit-vintage]/[runner:AbcTestSuite]/[test:AbcTest]/[dynamic:aaa 1(AbcTest)])
Finished:    AbcTestSuite ([engine:junit-vintage]/[runner:AbcTestSuite])
Finished:    JUnit Vintage ([engine:junit-vintage])
```

## Steps to reproduce

run `./gradlew test` with https://github.com/tomekgajdos/tmp-test-lib-a

## Context

 - Used versions (Jupiter/Vintage/Platform): 5.4.0
 - Build Tool/IDE: Gradle

## Deliverables

- [ ] Report intermediate containers as finished lazily, i.e. when a new test is started that is not a child of the current container or when the runner is finished (except for well-known runners such as `Parameterized`)

**Comment by marcphilipp** [comment]
Originally reported in https://github.com/gradle/gradle/issues/8787

**Comment by marcphilipp** [comment]
Fixed in #1820.
